core/src/main/scala/org/bitcoins/core/psbt/PSBT.scala

```diff
@@ -1695,23 +1732,34 @@ case class InputPSBTMap(elements: Vector[InputPSBTRecord])
       this
     } else {
       val newElements = {
-        val nonWitUtxoVec =
-          this.getRecords(NonWitnessUTXOKeyId)
-        val witScriptVec =
-          this.getRecords(WitnessScriptKeyId)
-        if (nonWitUtxoVec.size == 1 && witScriptVec.size == 1) {
-          val nonWitUtxo = nonWitUtxoVec.head.transactionSpent
+        if (nonWitnessOrUnknownUTXOOpt.isDefined) {
+          val nonWitUtxo = nonWitnessOrUnknownUTXOOpt.get.transactionSpent
           if (txIn.previousOutput.vout.toInt < nonWitUtxo.outputs.size) {
             val out = nonWitUtxo.outputs(txIn.previousOutput.vout.toInt)
-            elements
-              .filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
-              .filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
-              out)
+            out.scriptPubKey match {
+              case _: WitnessScriptPubKey =>
+                sortedElements
+                  .filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+                  .filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
+                  out)
+              case _: P2SHScriptPubKey =>
+                if (redeemScriptOpt.isDefined && redeemScriptOpt.get.redeemScript
+                      .isInstanceOf[WitnessScriptPubKey]) {
+                  sortedElements
+                    .filterNot(_.key.head == NonWitnessUTXOKeyId.byte)
+                    .filterNot(_.key.head == WitnessUTXOKeyId.byte) :+ WitnessUTXO(
+                    out)
```

Comment:
It seems that turning a `NonWitnessUTXO` into a `WitnessUTXO` should get its own method